File: go/generic/genFFT.go
package generic

import (
	"algos/helpers"
	"fmt"
	"math"
	"math/rand"
)

// GenFFT represents the FFT implementation
type GenFFT[N interface {
	IField[N]
	IMath[N]
	IPrimitiveRoots[N]
	ICopiable[N]
	IOrdered[N]
}] struct {
	c N
}

// NewGenFFT creates a new GenFFT instance
func NewGenFFT[N interface {
	IField[N]
	IMath[N]
	IPrimitiveRoots[N]
	ICopiable[N]
	IOrdered[N]
}](data N) GenFFT[N] {
	return GenFFT[N]{c: data}
}

// Transform performs the FFT on the given data
func (fft GenFFT[N]) Transform(data []N) {
	fft.transformInternal(data, -1)
}

// Inverse performs the inverse FFT on the given data
func (fft GenFFT[N]) Inverse(data []N) {
	fft.transformInternal(data, 1)
	n := len(data)
	norm := fft.c.coerceFromFloat(float64(n))
	for i := range data {
		data[i] = data[i].d(norm)
	}
}

// Test performs a round-trip FFT and inverse FFT, returning the RMS error
func (fft GenFFT[N]) Test(data []N) float64 {
	n := len(data)
	copyData := make([]N, n)
	for i, v := range data {
		copyData[i] = v.copy()
	}

	fft.Transform(data)
	fmt.Println("After transform:", data)
	fft.Inverse(data)
	fmt.Println("After inverse:", data)

	var diff float64
	for i := 0; i < n; i++ {
		realDiff := data[i].coerceToFloat() - copyData[i].coerceToFloat()
		diff += realDiff * realDiff
	}
	return math.Sqrt(diff / float64(n))
}

// MakeRandom generates random complex data
func (fft *GenFFT[N]) MakeRandom(n int) []ComplexField[N] {
	data := make([]ComplexField[N], n)
	for i := 0; i < n; i++ {
		data[i] = ComplexField[N]{
			Re: fft.c.coerceFromFloat(rand.Float64()),
			Im: fft.c.coerceFromFloat(rand.Float64()),
		}
	}
	return data
}

// transformInternal performs the FFT or inverse FFT
func (fft *GenFFT[N]) transformInternal(data []N, direction int) {
	n := len(data)
	if n <= 1 {
		return
	}
	logn := log2(n)

	bitreverse(data)

	roots := fft.c.precomputeRootsOfUnity(n, direction)
	//fmt.Println("Roots of unity:", roots)

	dual := 1
	for bit := 0; bit < logn; bit++ {
		for a := 0; a < dual; a++ {
			w := roots[a*(n/(2*dual))]
			for b := 0; b < n; b += 2 * dual {
				i := b + a
				j := b + a + dual

				wd := w.m(data[j])
				data[j] = data[i].s(wd)
				data[i] = data[i].a(wd)
			}
		}
		dual *= 2
	}
}

// log2 computes the base-2 logarithm of n, ensuring n is a power of 2
func log2(n int) int {
	log := 0
	k := 1
	for k < n {
		k *= 2
		log++
	}
	if n != (1 << log) {
		panic(fmt.Sprintf("FFT: Data length is not a power of 2!: %d", n))
	}
	return log
}

// bitreverse reorders the data array in bit-reversed order
func bitreverse[N any](data []N) {
	n := len(data)
	nm1 := n - 1
	i, j := 0, 0
	for i < nm1 {
		if i < j {
			data[i], data[j] = data[j], data[i]
		}
		k := n >> 1
		for k <= j {
			j -= k
			k >>= 1
		}
		j += k
		i++
	}
}

func TestGenFFT(n int, fieldType int) {
	var mode = 1 // let mode = 0 be for testing
	// fieldType: 1 = ComplexField[DoubleField], 2 = IntModP

	if mode != 0 {
		if n <= 0 {
			n = 16
		}
		rand := helpers.NewLCG(12345, 1345, 16645, 1013904)
		if fieldType == 1 {
			fmt.Println("Go Generic FFT Complex Field, ", n)
			varRandomNumbers := make([]ComplexField[DoubleField], n)
			for i := 0; i < n; i++ {
				varRandomNumbers[i] = ComplexField[DoubleField]{
					Re: DoubleField{Value: rand.NextDouble()},
					Im: DoubleField{Value: rand.NextDouble()},
				}
			}
			fft := NewGenFFT(ComplexField[DoubleField]{Re: DoubleField{Value: 3.0}, Im: DoubleField{Value: 4.0}})
			for j := 0; j < 10; j++ {
				fft.transformInternal(varRandomNumbers, -1)
				fft.transformInternal(varRandomNumbers, 1)
				fmt.Println("loop ", j, " done")
			}
		} else {
			fmt.Println("Go Generic FFT IntModP, ", n)
			varRandomNumbers := make([]IntModP, n)
			for i := 0; i < n; i++ {
				varRandomNumbers[i] = NewIntModP(uint64(rand.NextInt()))
			}
			fft := NewGenFFT(NewIntModP(3))
			for j := 0; j < 10; j++ {
				fft.transformInternal(varRandomNumbers, -1)
				fft.transformInternal(varRandomNumbers, 1)
				fmt.Println("loop ", j, " done")
			}

		}

	} else {
		rand := helpers.NewLCG(12345, 1345, 16645, 1013904)
		var randomNumbers [10]int
		var randomDoubles [10]float64
		for i := 0; i < 10; i++ {
			randomNumbers[i] = rand.NextInt()
			randomDoubles[i] = rand.NextDouble()
		}
		fmt.Println("Random Integers:", randomNumbers)
		fmt.Println("Random Doubles:", randomDoubles)

		c := ComplexField[DoubleField]{Re: DoubleField{Value: 3.0}, Im: DoubleField{Value: 4.0}}
		fft := NewGenFFT(c)
		n = 268435456

		data1 := make([]ComplexField[DoubleField], 0, n)
		data1 = append(data1, ComplexField[DoubleField]{
			Re: DoubleField{Value: 0.3618031071604718},
			Im: DoubleField{Value: 0.932993485288541},
		})
		data1 = append(data1, ComplexField[DoubleField]{
			Re: DoubleField{Value: 0.8330913489710237},
			Im: DoubleField{Value: 0.32647575623792624},
		})
		data1 = append(data1, ComplexField[DoubleField]{
			Re: DoubleField{Value: 0.2355237906476252},
			Im: DoubleField{Value: 0.34911535662488336},
		})
		data1 = append(data1, ComplexField[DoubleField]{
			Re: DoubleField{Value: 0.4480776326931518},
			Im: DoubleField{Value: 0.6381529437838686},
		})
		fmt.Println("Input data:", data1)
		fft.Transform(data1)
		fmt.Println("After FFT:", data1)
		fft.Inverse(data1)
		fmt.Println("After Inverse FFT:", data1)
		fmt.Printf("RMS Error: %.6f\n", fft.Test(data1))

		// Test with IntModP
		/*
			in1 := []uint64{38, 0, 44, 87, 6, 45, 22, 93, 0, 0, 0, 0, 0, 0, 0, 0}
			in2 := []uint64{80, 18, 62, 90, 17, 96, 27, 97, 0, 0, 0, 0, 0, 0, 0, 0}
			//out := []uint64{3040, 684, 5876, 11172, 5420, 16710, 12546, 20555, 16730, 15704, 21665, 5490, 13887, 4645, 9021, 0}
			prime := 40961
		*/

		in1 := []uint64{11400, 28374, 23152, 9576, 29511, 20787, 13067, 14015, 0, 0, 0, 0, 0, 0, 0, 0}
		in2 := []uint64{30268, 20788, 8033, 15446, 26275, 11619, 2494, 7016, 0, 0, 0, 0, 0, 0, 0, 0}
		//out := []uint64{ 345055200, 1095807432, 1382179648, 1175142886, 2016084656, 2555168834,2179032777, 1990011337, 1860865174, 1389799087, 942120918, 778961552,341270975, 126631482, 98329240, 0}
		prime := 3221225473
		SetModulus(uint64(prime))

		finite := NewIntModP(3)
		fftInt := NewGenFFT(finite)

		dataA := make([]IntModP, 0, len(in1))
		dataB := make([]IntModP, 0, len(in2))
		for i := 0; i < len(in1); i += 1 {
			dataA = append(dataA, NewIntModP(in1[i]))
			dataB = append(dataB, NewIntModP(in2[i]))
		}
		fmt.Println("Input data A:", dataA)
		fmt.Println("Input data B:", dataB)

		fftInt.Transform(dataA)
		fftInt.Transform(dataB)
		fmt.Println("After FFT A:", dataA)
		fmt.Println("After FFT B:", dataB)

		for i := range dataA {
			dataA[i] = dataA[i].m(dataB[i])
		}
		fmt.Println("After pointwise multiplication:", dataA)

		fftInt.Inverse(dataA)
		fmt.Println("After Inverse FFT:", dataA)
	}
}

File: typescript/generic/genFFT.ts
import { find_prime_congruent_one_mod_n } from '../helpers/find_prime';
import { LCG } from '../helpers/lcg';
import { IField } from './iField';
import { IMath } from './iMath';
import { IntModP } from './intModP';
import { IOrdered } from './iOrdered';
import { IPrimitiveRoots } from './iPrimitiveRoots';

// Class-based generic FFT, similar to the Rust GenFFT
export class GenFFT<C extends IField<C> & IOrdered<C> & IPrimitiveRoots<C> & IMath<C>> {
  c: C;

  constructor(data: C) {
    this.c = data;
  }

  transform(data: C[]): void {
    this.transformInternal(data, -1);
  }

  inverse(data: C[]): void {
    this.transformInternal(data, 1);
    const n = data.length;
    const norm = this.c.coerce(n);
    for (let i = 0; i < n; i++) {
      data[i].de(norm);
    }
  }

  test(data: C[]): number {
    const n = data.length;
    const copy = data.map(x => x.copy());

    //print before transform
    console.log("Before transform:", data.map(d => d.toString()).join(", "));

    this.transform(data);
    //print after transform
    console.log("After transform:", data.map(d => d.toString()).join(", "));

    this.inverse(data);
    //print after inverse
    console.log("After inverse:", data.map(d => d.toString()).join(", "));

    let diff = 0.0;
    for (let i = 0; i < n; i++) {
      const d = data[i];
      const orig = copy[i];
      const realDiff = d.coerce_to_number() - orig.coerce_to_number();
      diff += realDiff * realDiff;
    }
    return Math.sqrt(diff / n);
  }

  private transformInternal(data: C[], direction: number): void {
    const n = data.length;
    if (n === 0 || n === 1) return;

    this.bitreverse(data);

    const roots = this.c.precomputeRootsOfUnity(n, direction);

    let dual = 1;
    const logn = Math.log2(n);
    for (let bit = 0; bit < logn; bit++) {
      for (let a = 0; a < dual; a++) {
        const w = roots[a * (n / (2 * dual))];
        for (let b = 0; b < n; b += 2 * dual) {
          const i = b + a;
          const j = b + a + dual;
          const wd = w.m(data[j]);
          const u = data[i];
          data[j] = u.s(wd);
          data[i] = u.a(wd);
        }
      }
      dual *= 2;
    }
  }

  private bitreverse(data: C[]): void {
    const n = data.length;
    let j = 0;
    for (let i = 0; i < n; i++) {
      if (i < j) [data[i], data[j]] = [data[j], data[i]];
      let k = n >> 1;
      while (k && j >= k) {
        j -= k;
        k >>= 1;
      }
      j += k;
    }
  }
}

if(require.main == module) {
  // Usage: tsx genFFT.ts [n] [mode]
  // arg1 = size (N = power of 2)
  // arg2 = field type (0 = finite field, 1 = complex field)
  const args = process.argv.slice(2);
  const n = parseInt(args[0] || '16', 10);
  const mode = parseInt(args[1] || '0', 10);
  const rand = new LCG(12345, 1345, 16645, 1013904);

  const test = 1; // Set to 1 to run the test, 0 to just run the FFT without checking error
  if (!test) {
    if (mode === 1) {
      const { ComplexField } = require('./complexField');
      const { DoubleField } = require('./doubleField');
      const data = [];
      for (let i = 0; i < n; i++) {
        data.push(new ComplexField(new DoubleField(rand.nextDouble()), new DoubleField(rand.nextDouble())));
      }
      const fft = new GenFFT(data[0]);
      const error = fft.test(data);

      console.log(`FFT test completed with RMS error: ${error}`);
    } else {
      const data = [];
      const prime = find_prime_congruent_one_mod_n(n);
      IntModP.setModulus(prime);
      for (let i = 0; i < n; i++) {
        data.push(new IntModP(rand.nextInt() % prime));
      }
      const fft = new GenFFT(data[0]);
      const error = fft.test(data);
      console.log(`FFT test completed with RMS error: ${error}`);
    } 
  }
  else {
      const in1 = [38, 0, 44, 87, 6, 45, 22, 93, 0, 0, 0, 0, 0, 0, 0, 0];
      const in2 = [80, 18, 62, 90, 17, 96, 27, 97, 0, 0, 0, 0, 0, 0, 0, 0];
      //let out = [3040, 684, 5876, 11172, 5420, 16710, 12546, 20555, 16730, 15704, 21665, 5490, 13887, 4645, 9021, 0];
      let prime = 40961;
      IntModP.setModulus(prime);
      const data1 = [];
      const data2 = [];
      for (let i = 0; i < n; i++) {
        data1.push(new IntModP(in1[i]));
        data2.push(new IntModP(in2[i]));
      }
      const fft = new GenFFT(data1[0]);
      fft.transform(data1);
      fft.transform(data2);
      for (let i = 0; i < n; i++) {
        data1[i] = data1[i].m(data2[i]);
      }
      fft.inverse(data1);
      console.log("Result:", data1.map(d => d.toString()).join(", "));
           
  }
}
